dll/win32/shell32/wine/control.c

```diff
@@ -967,65 +960,201 @@ static	void	Control_DoLaunch(CPanel* panel, HWND hWnd, LPCWSTR wszCmd)
-        if (!aCPLPath)
+    /* Search for unquoted commas and spaces. */
+    for (i = 0; i < nLen; i++)
+    {
+        if (quoted && wszCmd[i] != L'"')
+            continue;
+        switch (wszCmd[i])
         {
-            aCPLPath = GlobalAddAtomW(applet->cmd);
+            case L'"':
+                quoted = !quoted;
+                break;
+            case L',':
+                if (!pchFirstComma) 
+                    pchFirstComma = &wszCmd[i];
+                else if (!pchSecondComma) 
+                    pchSecondComma = &wszCmd[i];
+                break;
+            case L' ': 
+                pchLastUnquotedSpace = &wszCmd[i];
+                break;
         }
+    }
 
-        aCPLName = GlobalFindAtomW(L"CPLName");
-        if (!aCPLName)
+    /* If no unquoted commas are found, parameters are either space separated, or the entire string
+     * is a CPL path. */
+    if (!pchFirstComma)
+    {
+        /* An unquoted space was found in the string. Assume the last word is the dialog box
+         * name/number. */
+        if (pchLastUnquotedSpace)
+        {
+            int nSpaces = 0;
+
+            while (pchLastUnquotedSpace[nSpaces] == L' ')
+                nSpaces++;
+
+            StringCchCopyNW(buffer, nLen + 1, wszCmd, pchLastUnquotedSpace - wszCmd);
+            StringCchCopyW(wszDialogBoxName, nLen + 1, pchLastUnquotedSpace + nSpaces);
+        }
+        /* No parameters were passed, the entire string is the CPL path. */
+        else
+        {
+            StringCchCopyW(buffer, nLen + 1, wszCmd);
+        }
+    }
+    /* If an unquoted comma was found, there are at least two parts of the string:
+     * - the CPL path
+     * - either a dialog box number preceeded by @, or a dialog box name.
+     * If there was a second unqoted comma, there is another part of the string:
+     * - the rest of the parameters. */
+    else
+    {
+        /* If there was no second unquoted comma in the string, the CPL path ends at thes
+          * null terminator. */
+        if (!pchSecondComma)
+            pchSecondComma = wszCmd + nLen;
+
+        StringCchCopyNW(buffer, nLen + 1, wszCmd, pchFirstComma - wszCmd);
+        StringCchCopyNW(wszDialogBoxName,
+                        nLen + 1,
+                        pchFirstComma + 1,
+                        pchSecondComma - pchFirstComma - 1);
+
+        if (pchSecondComma != wszCmd + nLen)
         {
-            aCPLName = GlobalAddAtomW(L"CPLName");
+            extraPmts = pchSecondComma + 1;
         }
+    }
+
+    /* Remove the quotes from both buffers. */
+    while ((ptr = StrChrW(buffer, '"')))
+        memmove(ptr, ptr+1, lstrlenW(ptr)*sizeof(WCHAR));
 
-        aCPLFlags = GlobalFindAtomW(L"CPLFlags");
-        if (!aCPLFlags)
+    while ((ptr = StrChrW(wszDialogBoxName, '"')))
+        memmove(ptr, ptr+1, lstrlenW(ptr)*sizeof(WCHAR));
+
+    if (wszDialogBoxName[0] == L'@')
+    {
+        sp = _wtoi(wszDialogBoxName + 1);
+    }
+
```

Comment:
This is needed to capture the 'extraPmts'
```
    if (wszDialogBoxName[0] != L'\0' && wszDialogBoxName[0] != L'@')
        extraPmts = wszDialogBoxName;
        
```
Needed to fix installation of mshtml on second phase boot.

Reply:
These changes seem to break some of Control_RunDLLW tests again. See my response to your comment on CORE-8981.